README.rst
=======
DryDock
=======

.. image:: https://badge.fury.io/py/drydock.png
    :target: http://badge.fury.io/py/drydock
    
.. image:: https://travis-ci.org/Nekroze/drydock.png?branch=master
    :target: https://travis-ci.org/Nekroze/drydock

.. image:: https://pypip.in/d/drydock/badge.png
    :target: https://pypi.python.org/pypi/drydock

A Docker cluster construction utility.

DryDock takes a simple (YAML_) specification file then can construct and
configure a cluster of Docker_ containers. DryDock will automatically
setup a reverse proxy, exposure of ports, and even persistent storage to
allow for easy future upgrading by simply rebuilding the DryDock
specification!

Features
--------

* Simple YAML_ configuration.
* Automatic Docker_ cluster provisioning/configuration
* Nginx_ reverse proxy configuration with HTTPS/SSL support
* Easy setup for persistent volumes.
* Share your DryDock specifications with the world.
* Container supervisor utilizing the Docker_ API.

TODO
----

* Provide a better update path, specific update commands, custom pre-post.
* Container monitoring, ensure a container is working else reconstruct it.
* Better support for stateless-ness and volumes
* Better unittest coverage.
* Ability to provide a custom docker images repository for master containers.
* An external facing DNS server or config generation.
* Control over a master container cluster, Drydock as a Service?.
* Web interface for control, logging and monitoring. Far future.

.. _Nginx: http://wiki.nginx.org/
.. _YAML: http://wikipedia.org/wiki/YAML
.. _Docker: https://www.docker.io/

HISTORY.rst
.. :changelog:

History
-------

0.6.11 (10-08-2014)
++++++++++++++++++

* Fixed: Improved environment variable handling

0.6.12 (19-06-2014)
++++++++++++++++++

* Changed: increased nginx upload limit for flexability

0.6.10 (19-06-2014)
++++++++++++++++++

* Changed: pass /etc/localtime to containers rather then /etc/timezone

0.6.9 (18-06-2014)
++++++++++++++++++

* Added: ``supervise`` now prints out each system command before calling it.

0.6.8 (14-06-2014)
++++++++++++++++++

* Fixed: ssl key generation commands now work and produce strong keys.

0.6.7 (14-06-2014)
++++++++++++++++++

* Removed: ``prepare`` no longer constructs base containers.

0.6.6 (14-04-2014)
++++++++++++++++++

* Fixed: Missing semicolon line ending for nginx https.

0.6.5 (14-04-2014)
++++++++++++++++++

* Added: Full options to start and supervise to allow for custom dns settings.

0.6.4 (31-03-2014)
++++++++++++++++++

* Changed: Altered the HTTPS/SSL reverse proxy config to hopefully work.

0.6.3 (23-03-2014)
++++++++++++++++++

* Fix: stale NFS file handles should no longer crash the supervisor.

0.6.2 (18-03-2014)
++++++++++++++++++

* Fix: remove conflicting args for containers ``-d`` and ``--rm``.

0.6.1 (17-03-2014)
++++++++++++++++++

* Fix: detect networking before running ``supervise``.

0.6.0 (17-03-2014)
++++++++++++++++++

* Changed: ``supervise`` now prepares and constructs all containers for a
**DryDock** cluster and will re-create them on failure.
* Changed: After running ``master`` two command will be provided,
one to pull all required images and first time setup,
another to run the **DryDock** cluster.

0.5.15 (17-03-2014)
+++++++++++++++++++

* Removed: got rid of master containers constructing special images.
* Changed: the master command just prepares the files and gives you a run
command for the master.
* Added: a volume map from the internal cluster containers based on FQDN and
the master containers ``/var/lib/{name}`` persistance directory.

0.5.14 (17-03-2014)
+++++++++++++++++++

* Changed: ``supervise`` will no longer reconstruct nginx at start.
* Changed: moved full master and cluster container construction to
construction of master image not first time run run.
* Changed: master images will now have ``-master`` appended to the end of
their name
* Changed: master images will now be named based on the specification name
not domain, as will their location in ``/var/lib/..``.
* Fix: Attempted to provide a better final run command for the master image.

0.5.13 (17-03-2014)
+++++++++++++++++++

* Changed: only construct master cluster when supervising.

0.5.12 (17-03-2014)
+++++++++++++++++++

* Added: new master container scripts to manage drydock cluster.

0.5.11 (16-03-2014)
+++++++++++++++++++

* Fix: disabled nginx event config.

0.5.10 (16-03-2014)
+++++++++++++++++++

* Fix: properly wrap final master command in qoutes for bash.

0.5.9 (16-03-2014)
++++++++++++++++++

* Fix: each image only gets pulled once via the pull command.
* Fix: more specific nginx config for use with skydock.

0.5.8 (16-03-2014)
++++++++++++++++++

* Changed: The shared data directory to map
``/var/lib/{domain}/drydock/data`` to the containers ``/mnt/data``.
* Added: Shared data container now works for master containers.

0.5.7 (16-03-2014)
++++++++++++++++++

* Changed: Moved master container prepare to first time run.

0.5.6 (16-03-2014)
++++++++++++++++++

* Fix: only create nginx configs where needed.

0.5.5 (16-03-2014)
++++++++++++++++++

* Changed: master containers only pull on creation and will construct at
first run.

0.5.4 (16-03-2014)
++++++++++++++++++

* Fix: volume mapped master containers drydock volume for persistance.

0.5.3 (16-03-2014)
++++++++++++++++++

* Added: display of final master container run command to end user.

0.5.2 (16-03-2014)
++++++++++++++++++

* Fix: master containers nginx sites are now read write.

0.5.1 (15-03-2014)
++++++++++++++++++

* Added: Customize your network interfaces, ips, and dns from cli.
* Fix: cleaned up template storage to allow manipulation.

0.5.0 (15-03-2014)
++++++++++++++++++

.. warning::
This release has a major emphasis on the ``master`` command and containers
and is released early to facilitate testing.

* Changed: ``master`` command now takes a specification to automatically
build it.
* Removed: any usage of ``supervisord`` in favor of the new
``drydock supervisor``.

0.4.3 (12-03-2014)
++++++++++++++++++

* Fix: use ``--name`` for naming containers for future proofing.
* Fix: use ``--dns`` for future proofing.

0.4.2 (11-03-2014)
++++++++++++++++++

* Fix: allow self connections to the host when nginx blocks external.

0.4.1 (09-03-2014)
++++++++++++++++++

* Fix: check for config files before removing them.
* Fix: ``supervise`` command will now recreate the nginx container each run.

0.4.0 (08-03-2014)
++++++++++++++++++


* Added: ``supervise`` command line command. DryDock has its own supervisor!
* Added: ``start`` and ``stop`` command line commands.
* Added: ``data`` in subcontainer specification maps volumes at
``/mnt/drydock``.
* Fix: All containers are passed their FQDN as their hostname
* Fix: ``pull`` command also grabs the containers required for the
``prepare`` command.
* Fix: Pass host timezone to subcontainers.
* Fix: volumes now go map to ``/var/lib/{domain}/{name}/``.

0.3.0 (28-02-2014)
++++++++++++++++++

* Added: reports at the end of running all the major commands.
* Added: ``envs`` to specification for environment variable definitions.
* Added: ``command`` to specification for run command definition.
* Added: ``pull`` command to download all images required for the
specification.
* Added ``specification`` to specification for external specification links.

0.2.0 (25-02-2014)
++++++++++++++++++

* Added: supervisor config writing is now an option.
* Added: deconstruct command to remove a specification.

0.1.0 (25-02-2014)
++++++++++++++++++

* First release on PyPI.


test/test_duster.py
"""
Tests for `drydock.duster` module.
"""
import pytest
from drydock.duster import Container, MetaContainer
import yaml


class TestContainer(object):
    config = """---
name: blog
base: nekroze/wordpress
domain: nekroze.com
exposed_ports:
    22: 22
    2222: 222
http_port: 8081
https_port: 4431
external: No
..."""

    def test_construction(self):
        container = Container(**yaml.load(self.config))
        assert container.name == "blog"
        assert container.base == "nekroze/wordpress"
        assert container.domain == "nekroze.com"
        assert container.exposed_ports == {22:22, 2222:222}
        assert container.http_port == 8081
        assert container.https_port == 4431
        assert container.external is False
        assert container.skyfqdn == "blog.wordpress.containers.drydock"
        assert container.fqdn == "blog.nekroze.com"

    def test_supervisord(self):
        container = Container(**yaml.load(self.config))
        assert container.get_supervisor_config() == """[program:blog]
command=docker start blog
autostart=true
autorestart=true"""

    def test_nginx(self):
        container = Container(**yaml.load(self.config))
        expected = """upstream blog {
    server blog.wordpress.containers.drydock;
}
server {
    listen       80;
    server_name  blog.nekroze.com;

    deny all;
    allow 192.168.1.0/24;
    allow 192.168.0.0/24;

    access_log  /var/log/nginx/log/blog.nekroze.com.access.log  main;
    error_log  /var/log/nginx/log/blog.nekroze.com.error.log;

    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_redirect off;
        proxy_buffering off;

        proxy_pass https://blog:8081/;
    }
}
server {
    listen 443;
    server_name blog.nekroze.com;

    deny all;
    allow 192.168.1.0/24;
    allow 192.168.0.0/24;

    access_log  /var/log/nginx/log/blog.nekroze.com.access.log  main;
    error_log  /var/log/nginx/log/blog.nekroze.com.error.log;

    ssl on;
    ssl_session_timeout 5m;
    ssl_protocols SSLv2 SSLv3 TLSv1;
    ssl_ciphers ALL:!ADH:!EXPORT56:RC4+RSA:+HIGH:+MEDIUM:+LOW:+SSLv2:+EXP;
    ssl_prefer_server_ciphers on;
    ssl_certificate /etc/nginx/certs/server.crt;
    ssl_certificate_key /etc/nginx/certs/server.key;

    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_redirect off;
        proxy_buffering off;

        proxy_pass https://blog:4431/;
    }
}"""
        assert container.get_nginx_config() == expected

test/test_drydock.py
"""
Tests for `drydock` module.
"""
import pytest
from drydock import drydock


class TestDrydock(object):

    @classmethod
    def setup_class(cls):
        pass

    def test_something(self):
        pass

    @classmethod
    def teardown_class(cls):
        pass

test/test_container.py
"""
Tests for `drydock.duster` module and Container object.
"""
import pytest
from drydock.duster import Container
from drydock.templates import NETWORK
import yaml


class TestContainer(object):
    config = """
name: blog
base: nekroze/wordpress
domain: nekroze.com
command: crump
envs:
    DB: 123
exposed_ports:
    22: 22
    2222: 222
http_port: 8081
https_port: 4431
external: No
"""

    def test_construction(self):
        container = Container(**yaml.load(self.config))
        assert container.name == "blog"
        assert container.base == "nekroze/wordpress"
        assert container.domain == "nekroze.com"
        assert container.exposed_ports == {22:22, 2222:222}
        assert container.http_port == 8081
        assert container.https_port == 4431
        assert container.external is False
        assert container.skyfqdn == "blog.wordpress.containers.drydock"
        assert container.fqdn == "blog.nekroze.com"

    def test_nginx(self):
        container = Container(**yaml.load(self.config))
        NETWORK["dns"] = "172.17.42.1"
        container.http = True
        container.https = True
        expected = """server {
    listen       80;
    server_name  blog.nekroze.com;

    access_log  /var/log/nginx/blog.nekroze.com.access.log  combined;
    error_log  /var/log/nginx/blog.nekroze.com.error.log;

    client_max_body_size 500M;

    resolver 172.17.42.1 valid=5s;
    resolver_timeout 5s;

    location / {
        deny    {gateway};
        allow   {lan}/24;
        allow   {docker}/24;
        deny    all;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_redirect off;
        proxy_buffering off;

        proxy_pass http://blog.wordpress.containers.drydock:8081/;
    }
}
server {
    listen       443;
    server_name  blog.nekroze.com;

    access_log  /var/log/nginx/blog.nekroze.com.access.log  combined;
    error_log  /var/log/nginx/blog.nekroze.com.error.log;

    client_max_body_size 500M;

    resolver 172.17.42.1 valid=5s;
    resolver_timeout 5s;

    ssl                         on;
    ssl_session_timeout         10m;
    ssl_protocols               SSLv3 TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers                 RC4:HIGH:!aNULL:!MD5;
    ssl_prefer_server_ciphers   on;
    ssl_certificate /etc/nginx/certs/server.crt;
    ssl_certificate_key /etc/nginx/certs/server.key;

    location / {
        deny    {gateway};
        allow   {lan}/24;
        allow   {docker}/24;
        deny    all;

        proxy_next_upstream error timeout invalid_header http_500 http_502 http_503 http_504;
        proxy_set_header        Accept-Encoding   "";
        proxy_set_header        Host            $host;
        proxy_set_header        X-Real-IP       $remote_addr;
        proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header        X-Forwarded-Proto $scheme;
        add_header              Front-End-Https   on;
        proxy_redirect off;

        proxy_pass http://blog.wordpress.containers.drydock:4431/;
    }
}"""
        assert container.get_nginx_config() == expected

    def test_container_commands(self):
        container = Container(**yaml.load(self.config))

        assert container.get_docker_command() == \
               "docker run -d --dns {dockerdns} --name blog -h blog.nekroze.com -p 22:22 -p 2222:222 -v /etc/localtime:/etc/localtime:ro -e DB=\"123\" nekroze/wordpress crump"


drydock/report.py
"""End of run reports."""
__author__ = 'Taylor \"Nekroze\" Lawson'
import sys


class Report(object):
    def __init__(self):
        self.containers = {"success": {}, "failed": {}}
        self.paths = []
        self.commands = {"success": {}, "failed": {}}

    def exit(self):
        """Exit the program with the appropriate code based on the report."""
        if self.containers["failed"] or self.commands["failed"]:
            sys.exit(1)
        else:
            sys.exit(0)

    def container(self, name, command, code):
        """Report on the status of a container"""
        key = "success" if code == 0 else "failed"
        self.containers[key][name] = command

    def path(self, path):
        """Report on the status of a path"""
        self.paths.append(path)

    def command(self, description, command, code):
        """Report on the status of a command"""
        key = "success" if code == 0 else "failed"
        self.commands[key][description] = command

    def render(self, remove=False):
        """
        Render the current state of this report.

        If the ``remove`` argument is true then a removal report will be
        rendered.
        """
        output = ["DryDock Report:"]

        success = len(self.containers["success"].keys())
        failed = len(self.containers["failed"].keys())
        if success:
            output.append("    {} container(s) {}.".format(
                success, "deleted" if remove else "created"))
            output.append("    >" + ', '.join(
                name for name in self.containers["success"].keys()))
        if failed:
            output.append("    {} container(s) failed to be {}.".format(
                failed, "deleted" if remove else "created"))
            for name, command in self.containers["failed"].items():
                output.append("    >{}: {}".format(name, command))

        if success or failed:
            output.append("")

        success = len(self.commands["success"].keys())
        failed = len(self.commands["failed"].keys())
        if success:
            output.append("    {} command(s) succeeded.".format(success))
        if failed:
            output.append("    {} command(s) failed.".format(failed))
            for desc, command in self.commands["failed"].items():
                output.append("    >{}: {}".format(desc, command))

        if success or failed:
            output.append("")

        if len(self.paths):
            output.append("    {} path(s) {}.".format(
                len(self.paths), "deleted" if remove else "written"))
            for path in self.paths:
                output.append("    >" + path)

        return '\n'.join(output)

drydock/drydock.py
from __future__ import print_function
from . import __version__
__doc__ = """DryDock v{}

Usage:
    drydock construct [options] <specification>
    drydock deconstruct <specification>
    drydock pull <specification>
    drydock start [options] <specification>
    drydock supervise [options] <specification>
    drydock stop <specification>
    drydock prepare [options]
    drydock master [options] <specification>
    drydock --help | --version

Options:
    -n --dns=<IP>           External dns server to use. [default: 8.8.8.8]
    -d --docker=<IFACE>     Docker  network interface. [default: docker0]
    -l --lan=<IFACE>        Local area network interface. [default: eth0]
    -g --gate=<INT>         Position of network gateway. [default: 1]

    -s --ssl                Generate SSL certificates.

    -h --help               Show this screen.
    -v --version            Show current version.

Commands:
    construct    Construct the given specification.
    deconstruct  Deconstruct the given specification.
    prepare      Prepare docker for constructing a drydock specification.
    master       Create a container that can run a drydock specification.

DryDock takes a simple (YAML) specification file then can construct and
configure a cluster of docker containers. DryDock will automatically setup a
reverse proxy, exposure of ports, and even persistent storage to allow for
easy future upgrading by simply rebuilding the DryDock specification!

For documentation go to http://dry-dock.readthedocs.org/
""".format(__version__)
import yaml
from . import construction
from .duster import MetaContainer
from docopt import docopt
from .templates import prepare_networking


def main():
    """Main entry point."""
    args = docopt(__doc__, version="DryDock v" + __version__)

    if args["construct"]:
        prepare_networking(args["--lan"], args["--docker"], args["--gate"],
                           args["--dns"])
        with open(args["<specification>"], 'r') as drydock:
            construction.construct(
                MetaContainer(**yaml.load(drydock.read())))

    elif args["deconstruct"]:
        with open(args["<specification>"], 'r') as drydock:
            construction.deconstruct(
                MetaContainer(**yaml.load(drydock.read())))

    elif args["pull"]:
        with open(args["<specification>"], 'r') as drydock:
            construction.pull(
                MetaContainer(**yaml.load(drydock.read())))

    elif args["start"]:
        prepare_networking(args["--lan"], args["--docker"], args["--gate"],
                           args["--dns"])
        with open(args["<specification>"], 'r') as drydock:
            construction.start(
                MetaContainer(**yaml.load(drydock.read())))

    elif args["supervise"]:
        prepare_networking(args["--lan"], args["--docker"], args["--gate"],
                           args["--dns"])
        with open(args["<specification>"], 'r') as drydock:
            construction.supervise(
                MetaContainer(**yaml.load(drydock.read())))

    elif args["stop"]:
        with open(args["<specification>"], 'r') as drydock:
            construction.stop(
                MetaContainer(**yaml.load(drydock.read())))

    elif args["master"]:
        prepare_networking(args["--lan"], args["--docker"], args["--gate"],
                           args["--dns"])
        with open(args["<specification>"], 'r') as drydock:
            construction.master(
                MetaContainer(**yaml.load(drydock.read())),
                args["<specification>"])

    elif args["prepare"]:
        prepare_networking(args["--lan"], args["--docker"], args["--gate"],
                           args["--dns"])
        construction.prepare(args["--ssl"])


if __name__ == "__main__":
    main()


drydock/templates.py
"""A collection of templates and rendering functions."""
import subprocess


def get_ip_range(interface="docker0"):
    cmd = "ip address show dev " + interface
    addr = subprocess.check_output(cmd.split()).split()
    ip = addr[addr.index('inet') + 1].split('/')[0]
    return ip.split('.')[:-1] + ['0']


TEMPLATES = {"NGINX": {}, "BASE": {}}

TEMPLATES["BASE"]["CONTAINERS"] = """docker run -d -p {dockerdns}:53:53/udp --name skydns crosbymichael/skydns -nameserver {dns}:53 -domain drydock
docker run -d -v /var/run/docker.sock:/docker.sock --name skydock --dns {dockerdns} --link skydns:skydns crosbymichael/skydock -ttl 30 -environment containers -s /docker.sock -domain drydock
docker run -d -p 80:80 -p 443:443 --name nginx --dns {dockerdns} -v /etc/nginx/certs:/etc/nginx/certs -v /etc/nginx/sites-enabled:/etc/nginx/sites-enabled -v /var/log/nginx:/var/log/nginx dockerfile/nginx
cd /etc/nginx/certs && openssl genrsa -out server.key 2048 && openssl req -new -key server.key -out server.csr && openssl x509 -req -days 365 -in server.csr -signkey server.key -out server.crt"""

TEMPLATES["NGINX"]["HEADER"] = """events {
    worker_connections 1024;
    use epoll;
}"""

TEMPLATES["NGINX"]["UPSTREAM"] = """upstream {name} {{
    server {skyfqdn};
}}"""

TEMPLATES["NGINX"]["HTTP"] = """server {{
    listen       80;
    server_name  {fqdn};

    access_log  /var/log/nginx/{fqdn}.access.log  combined;
    error_log  /var/log/nginx/{fqdn}.error.log;

    client_max_body_size 500M;

    resolver {dns} valid=5s;
    resolver_timeout 5s;

    location / {{
        {rules}
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-NginX-Proxy true;
        proxy_redirect off;
        proxy_buffering off;

        proxy_pass http://{skyfqdn}:{port}/;
    }}
}}"""

TEMPLATES["NGINX"]["HTTPS"] = """server {{
    listen       443;
    server_name  {fqdn};

    access_log  /var/log/nginx/{fqdn}.access.log  combined;
    error_log  /var/log/nginx/{fqdn}.error.log;

    resolver {dns} valid=5s;
    resolver_timeout 5s;

    client_max_body_size 500M;

    ssl                         on;
    ssl_session_timeout         10m;
    ssl_protocols               SSLv3 TLSv1 TLSv1.1 TLSv1.2;
    ssl_ciphers                 RC4:HIGH:!aNULL:!MD5;
    ssl_prefer_server_ciphers   on;
    ssl_certificate /etc/nginx/certs/server.crt;
    ssl_certificate_key /etc/nginx/certs/server.key;

    location / {{
        {rules}
        proxy_next_upstream error timeout invalid_header http_500 http_502 http_503 http_504;
        proxy_set_header        Accept-Encoding   "";
        proxy_set_header        Host            $host;
        proxy_set_header        X-Real-IP       $remote_addr;
        proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header        X-Forwarded-Proto $scheme;
        add_header              Front-End-Https   on;
        proxy_redirect off;

        proxy_pass http://{skyfqdn}:{port}/;
    }}
}}"""


TEMPLATES["NGINX"]["RULES"] = """deny    {gateway};
        allow   {lan}/24;
        allow   {docker}/24;
        deny    all;
"""


NETWORK = {}


def prepare_networking(lani="eth0", dockeri="docker0",
                       gateway='1', dns="8.8.8.8"):
    """Stores useful network information and renders templates."""
    lanip = get_ip_range(lani)
    lan = '.'.join(lanip)
    gateway = '.'.join(lanip[:-1] + [gateway])
    dockerip = get_ip_range(dockeri)
    docker = '.'.join(dockerip)
    dockerdns = '.'.join(dockerip[:-1] + ['1'])

    TEMPLATES["NGINX"]["RULES"] = TEMPLATES["NGINX"]["RULES"].format(
        lan=lan, gateway=gateway, docker=docker)

    TEMPLATES["BASE"]["CONTAINERS"] = TEMPLATES["BASE"]["CONTAINERS"].format(
        lan=lan, gateway=gateway, docker=docker, dns=dns, dockerdns=dockerdns)
    NETWORK["lan"] = lan
    NETWORK["gateway"] = gateway
    NETWORK["docker"] = docker
    NETWORK["dockerdns"] = dockerdns
    NETWORK["dns"] = dns


def render_nginx_config(container):
    """
    Render the nginx reverse proxy configuration for the given container.
    """
    if not container.http_port and not container.https_port:
        return ""

    config = []

    if not container.external:
        rules = TEMPLATES["NGINX"]["RULES"]
    else:
        rules = ""

    if container.http:
        config.append(TEMPLATES["NGINX"]["HTTP"].format(
            skyfqdn=container.skyfqdn, name=container.name, dns=NETWORK["dns"],
            port=container.http_port, fqdn=container.fqdn, rules=rules))

    if container.https:
        config.append(TEMPLATES["NGINX"]["HTTPS"].format(
            skyfqdn=container.skyfqdn, name=container.name, dns=NETWORK["dns"],
            port=container.https_port, fqdn=container.fqdn, rules=rules))

    return '\n'.join(config)


def base_commands():
    """Return the base commands in a list"""
    return [cmd for cmd in TEMPLATES["BASE"]["CONTAINERS"].split('\n')]


drydock/duster.py
"""DryDock cluster specification."""
from .templates import render_nginx_config, TEMPLATES
from six.moves.urllib.request import urlopen
import yaml


class Container(object):
    """A docker container specification."""
    def __init__(self, name, base="ubuntu", exposed_ports=None, external=True,
                 http_port="80", https_port="443", domain="", volumes=None,
                 envs=None, command=None, data=False):
        self.name = name
        self.domain = ""
        self.fqdn = ""
        self.base = base
        self.envs = envs if envs else {}
        self.data = data
        self.exposed_ports = {}
        self.command = command
        self.http = False
        self.https = False
        if exposed_ports:
            self.exposed_ports.update(exposed_ports)
        self.http_port = http_port
        self.https_port = https_port
        self.external = external
        self.volumes = volumes if volumes else []
        self.set_domain(domain)

    def set_domain(self, domain):
        """Set fqdn and domain."""
        self.domain = domain
        if self.name == "root":
            self.fqdn = self.domain
        else:
            self.fqdn = self.name + '.' + self.domain
        self.skyfqdn = '.'.join([self.name, self.base.split('/')[-1],
                                 "containers", "drydock"])

    def get_portmaps(self):
        """Get docker command port maps."""
        output = []
        for external in sorted(self.exposed_ports.keys()):
            output.append("-p {0}:{1}".format(external,
                                           self.exposed_ports[external]))
        return output

    def get_volumemaps(self):
        """Return docker command volume maps."""
        output = []
        for path in self.volumes:
            output.append("-v /var/lib/{0}/{1}{2}:{2}".format(
                self.domain, self.name, path))
        return output

    def get_envs(self):
        """Return docker environment variable arguments."""
        output = []
        for name, value in self.envs.items():
            output.append("-e '{0}={1}'".format(name, value))
        return output

    def get_docker_command(self):
        """Return the docker command required to create this container."""
        cmd = ["docker run -d --dns {dockerdns}"]
        cmd.append("--name " + self.name)
        cmd.append("-h " + self.fqdn)
        cmd.extend(self.get_portmaps())
        if self.data:
            cmd.append("-v /var/lib/{}/drydock/data:/mnt/data".format(
                self.domain))
        cmd.append("-v /etc/localtime:/etc/localtime:ro")
        cmd.extend(self.get_volumemaps())
        cmd.extend(self.get_envs())
        cmd.append(self.base)
        if self.command:
            cmd.append(self.command)
        return ' '.join(cmd)

    def get_nginx_config(self):
        """Get a rendered nginx configuration file for this container."""
        return render_nginx_config(self)


class MetaContainer(Container):
    """A container that stores containers."""
    def __init__(self, name, domain, subcontainers, base="nekroze/drydock", *args,  **kwargs):
        super(MetaContainer, self).__init__(*args, name=name, base=base,
                                            domain=domain, **kwargs)
        if not self.command:
            self.command = \
                "startdocker && drydock supervise /drydock/specification.yaml"
        self.containers = {}
        self.reverse_proxies = {}
        self.fqdn = domain
        self.exposed_ports[self.http_port] = "80"
        self.exposed_ports[self.https_port] = "443"

        for sub in subcontainers:
            if "specification" in sub:
                sub = self.grab_specification(sub["specification"], sub)
            container = Container(**sub)
            container.http = "http_port" in sub
            container.https = "https_port" in sub
            self.add_container(container)

    def grab_specification(self, link, base):
        """Create a container specification based off of a web link."""
        spec = yaml.load(urlopen(link).read()) if link else {}
        spec.update(base)
        del spec["specification"]
        return spec

    def set_domain(self, domain):
        """Set fqdn and domain to the same value."""
        self.domain = domain
        self.fqdn = domain

    def add_container(self, container):
        """Add the given container object to this meta container."""
        if not container.domain and self.domain:
            container.set_domain(self.domain)

        self.containers[container.name] = container
        self.exposed_ports.update(container.exposed_ports)

        fqdn = container.name + '.' + self.domain
        self.reverse_proxies[fqdn] = {
            "skyfqdn": container.skyfqdn,
            "external": container.external,
            "http": container.http_port,
            "https": container.https_port
        }

    def get_nginx_config(self):
        """Get a rendered nginx configuration file for all containers."""
        #output = [TEMPLATES["NGINX"]["HEADER"]]
        output = []
        for name in sorted(self.containers.keys()):
            container = self.containers[name]
            output.append(container.get_nginx_config())
        return '\n'.join(output)

    def get_volumemaps(self):
        """Return docker command volume maps."""
        output = []
        for path in self.volumes:
            output.append("-v /var/lib/{0}{1}:{1}".format(self.name, path))
        return output

    def get_docker_command(self):
        """
        Return the docker command required to create specification in
        a master container.
        """
        cmd = ["docker run --privileged"]
        cmd.append("--name {}".format(self.name))
        cmd.append("-h " + self.fqdn)
        cmd.extend(self.get_portmaps())
        ngx = "-v /var/lib/{}/nginx/sites-enabled:/etc/nginx/sites-enabled"
        cmd.append(ngx.format(self.fqdn))
        ngx = "-v /var/lib/{}/nginx/ssl:/etc/nginx/ssl:ro"
        cmd.append(ngx.format(self.fqdn))
        cmd.append("-v /etc/timezone:/etc/timezone:ro")
        cmd.append("-v /var/lib/{0}/docker:/var/lib/docker".format(
            self.name))
        cmd.append("-v /var/lib/{0}:/var/lib/{1}".format(
            self.name, self.fqdn))
        cmd.append("-v /var/lib/{0}/drydock/data:/var/lib/{0}/drydock/data".format(
            self.name))
        cmd.append("-v /var/lib/{0}/drydock:/drydock:ro".format(self.fqdn))
        cmd.extend(self.get_volumemaps())
        cmd.append(self.base)
        return ' '.join(cmd)
